File: LeetCode/problems/p0007/java/src/p007/java/P007.java
//[COMPLETED]

package p007;

public class P007 {

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println("123: " + s.reverse(123));
        System.out.println("-123: " + s.reverse(-123));
        System.out.println("120: " + s.reverse(120));
        System.out.println("0: " + s.reverse(0));
        System.out.println("0: " + s.reverse(1));
        System.out.println("1534236469: " + s.reverse(1534236469));
        System.out.println("-2147483648: " + s.reverse(-2147483648));
        
        int x = -2147483648;
        int y = Math.abs(-2147483648);
        System.out.println(x);
        System.out.println(y);
        System.out.println(x-1);
    }
    
private static class Solution {
    public int reverse(int x) {

        long absNum = Math.abs((long) x);
        String str = "" + absNum;

        String reversed = "";
        while (str != null && str.length() > 0) {
            reversed = reversed + str.charAt(str.length() - 1);
            str = str.substring(0, str.length() -1);
        }

        while (reversed.length() > 0 && reversed.charAt(0) == '0')
            reversed = reversed.substring(1);

        if (x < 0)
            reversed = '-' + reversed;

        if (reversed == null || reversed.equals(""))
            return 0;
        
        long ans = Long.parseLong(reversed);
        if (ans > Integer.MAX_VALUE || ans < Integer.MIN_VALUE)
            return 0;
        else 
            return (int) ans;
    }
}

}

File: leet-code/src/problems/p0007/python/p007.py
#[COMPLETED]

class Solution(object):
    """
    :type x: int
    :rtype: int
    """
    def reverse(self, x):
        s = str(x)
        neg = False
        if s[0] == '-':
            s = s[1:]
            neg = True
        s = s[::-1] if neg is False else '-' + s[::-1]
        x = int(s)
        if x < -2**31 or x > 2**31 - 1:
            return 0
        return x

sol = Solution()
print(sol.reverse(123), "expected 321")
print(sol.reverse(-123), "expected -321")
print(sol.reverse(120), "expected 21")
print(sol.reverse(1534236469), "expected 0")
print(sol.reverse(1563847412), "expected 0")

print(-2**31)
print(2**31 - 1)
